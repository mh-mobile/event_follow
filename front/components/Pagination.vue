<template lang="pug">
.event_pagination
  ul
    li(v-for="(page, index) in pages" :key="index" @click="showPage(page.value)", :class="page.class")
      | {{ page.label }}
</template>

<script lang="ts">
import { defineComponent, computed } from "@nuxtjs/composition-api"

export default defineComponent({
  props: {
    totalPages: {
      type: Number,
      default: 0
    },
    currentPage: {
      type: Number,
      default: 1
    },
    pageWindow: {
      type: Number,
      default: 2
    }
  },
  setup(props, { root }) {
    if (process.server) {
      return {
        pages: [],
        showPage: null
      }
    }

    const pages = computed(() => {
      const currentPage = props.currentPage ? props.currentPage : 1
      const totalPages = props.totalPages ? props.totalPages : 0
      const pageWindow = props.pageWindow ? props.pageWindow : 2
      let _pages = []
      _pages.push({
        label: "最初",
        value: "1",
        class: currentPage === 1 ? "click_disable" : ""
      })

      _pages.push({
        label: "前へ",
        value: currentPage - 1,
        class: currentPage === 1 ? "click_disable" : ""
      })

      if (currentPage > 1 + pageWindow) {
        _pages.push({
          label: "...",
          value: "",
          class: "intermediate_page click_disable"
        })
      }

      for (let page = currentPage - pageWindow; page < currentPage; page++) {
        if (page < 1) continue

        _pages.push({
          label: page,
          value: page,
          class: "page_item"
        })
      }

      _pages.push({
        label: currentPage,
        value: currentPage,
        class: "current_page click_disable page_item"
      })

      for (
        let page = currentPage + 1;
        page <= currentPage + pageWindow;
        page++
      ) {
        if (page > totalPages) continue

        _pages.push({
          label: page,
          value: page,
          class: "page_item"
        })
      }

      if (currentPage < totalPages - pageWindow) {
        _pages.push({
          label: "...",
          value: "",
          class: "intermediate_page click_disable"
        })
      }

      _pages.push({
        label: "次へ",
        value: currentPage + 1,
        class: currentPage === totalPages ? "click_disable" : ""
      })

      _pages.push({
        label: "最後",
        value: totalPages,
        class: currentPage === totalPages ? "click_disable" : ""
      })

      return _pages
    })

    const showPage = (page: number) => {
      root.$router.replace(`/events?page=${page}`)
    }

    return {
      pages: pages,
      showPage: showPage
    }
  }
})
</script>

<style lang="scss" scoped>
@media all and (min-width: 480px) {
  .event_pagination {
    width: 100%;
  }

  .event_pagination ul {
    display: flex;
    flex-direction: row;
    justify-content: center;

    li {
      padding: 10px 15px;
      border-top: 1px solid #ccc;
      border-bottom: 1px solid #ccc;
      cursor: pointer;
      display: flex;
      flex-direction: row;
      justify-content: center;
      align-items: center;
      &:not(:last-child) {
        border-right: 1px solid #ccc;
      }

      &:first-child {
        border-left: 1px solid #ccc;
        border-top-left-radius: 5px;
        border-bottom-left-radius: 5px;
      }

      &:last-child {
        border-right: 1px solid #ccc;
        border-top-right-radius: 5px;
        border-bottom-right-radius: 5px;
      }
    }

    .page_item {
      min-width: 50px;
    }

    .current_page {
      background-color: #4d4f53 !important;
      color: #fff !important;
      border-top: 1px solid #4d4f53 !important;
      border-bottom: 1px solid #4d4f53 !important;
    }

    .intermediate_page {
      border-top: 1px solid #ccc !important;
      border-bottom: 1px solid #ccc !important;
      color: #000 !important;
      background: #fff !important;
      min-width: 50px;
    }

    .click_disable {
      pointer-events: none;
      color: #ccc;
    }
  }
}

@media all and (max-width: 480px) {
  .event_pagination {
    width: 100%;
  }

  .event_pagination ul {
    width: 100%;
    display: flex;
    flex-direction: row;
    justify-content: stretch;
    font-size: 0.8em; 

    li {
      flex-grow: 1;
      height: 40px;
      border-top: 1px solid #ccc;
      border-bottom: 1px solid #ccc;
      background-color: #ffffff;
      cursor: pointer;
      display: flex;
      flex-direction: row;
      justify-content: center;
      align-items: center;
      white-space: nowrap;
      
      &:not(:last-child) {
        border-right: 1px solid #ccc;
      }

      &:first-child {
        border-left: 1px solid #ccc;
        border-top-left-radius: 5px;
        border-bottom-left-radius: 5px;
      }

      &:last-child {
        border-right: 1px solid #ccc;
        border-top-right-radius: 5px;
        border-bottom-right-radius: 5px;
      }
    }

    .page_item {
      min-width: 50px;
    }

    .current_page {
      background-color: #4d4f53 !important;
      color: #fff !important;
      border-bottom: 1px solid #ccc;
      background-color: #ffffff;
    }

    .intermediate_page {
      border-top: 1px solid #ccc !important;
      border-bottom: 1px solid #ccc !important;
      color: #000 !important;
      background: #fff !important;
      min-width: 50px;
    }

    .click_disable {
      pointer-events: none;
      color: #ccc;
    }
  }
}
</style>
