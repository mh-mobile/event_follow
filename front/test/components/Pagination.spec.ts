import { createLocalVue, shallowMount } from "@vue/test-utils"
import VueCompositionApi from "@vue/composition-api"
import Pagination from "@/components/Pagination.vue"

const localVue = createLocalVue()
localVue.use(VueCompositionApi)

let totalPages: number
let pageWindow: number

describe("Pagination.vue", () => {
  beforeEach(() => {
    totalPages = 100
    pageWindow = 2
  })

  it("currentPage=2", () => {
    const wrapper = shallowMount(Pagination, {
      localVue,
      propsData: {
        totalPages,
        pageWindow,
        currentPage: 2
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("currentPage=3", () => {
    const wrapper = shallowMount(Pagination, {
      localVue,
      propsData: {
        totalPages,
        pageWindow,
        currentPage: 3
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("currentPage=4", () => {
    const wrapper = shallowMount(Pagination, {
      localVue,
      propsData: {
        totalPages,
        pageWindow,
        currentPage: 4
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("currentPage=10", () => {
    const wrapper = shallowMount(Pagination, {
      localVue,
      propsData: {
        totalPages,
        pageWindow,
        currentPage: 10
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("currentPage=97", () => {
    const wrapper = shallowMount(Pagination, {
      localVue,
      propsData: {
        totalPages,
        pageWindow,
        currentPage: 97
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("currentPage=98", () => {
    const wrapper = shallowMount(Pagination, {
      localVue,
      propsData: {
        totalPages,
        pageWindow,
        currentPage: 98
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("currentPage=99", () => {
    const wrapper = shallowMount(Pagination, {
      localVue,
      propsData: {
        totalPages,
        pageWindow,
        currentPage: 99
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("currentPage=100", () => {
    const wrapper = shallowMount(Pagination, {
      localVue,
      propsData: {
        totalPages,
        pageWindow,
        currentPage: 100
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
