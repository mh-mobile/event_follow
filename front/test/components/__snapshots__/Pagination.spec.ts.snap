// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination.vue currentPage=2 1`] = `
"<div class=\\"event_pagination\\">
  <ul>
    <li class=\\"\\">最初</li>
    <li class=\\"\\">前へ</li>
    <li class=\\"page_item\\">1</li>
    <li class=\\"current_page click_disable page_item\\">2</li>
    <li class=\\"page_item\\">3</li>
    <li class=\\"page_item\\">4</li>
    <li class=\\"\\">次へ</li>
    <li class=\\"\\">最後</li>
  </ul>
</div>"
`;

exports[`Pagination.vue currentPage=3 1`] = `
"<div class=\\"event_pagination\\">
  <ul>
    <li class=\\"\\">最初</li>
    <li class=\\"\\">前へ</li>
    <li class=\\"page_item\\">1</li>
    <li class=\\"page_item\\">2</li>
    <li class=\\"current_page click_disable page_item\\">3</li>
    <li class=\\"page_item\\">4</li>
    <li class=\\"page_item\\">5</li>
    <li class=\\"\\">次へ</li>
    <li class=\\"\\">最後</li>
  </ul>
</div>"
`;

exports[`Pagination.vue currentPage=4 1`] = `
"<div class=\\"event_pagination\\">
  <ul>
    <li class=\\"\\">最初</li>
    <li class=\\"\\">前へ</li>
    <li class=\\"page_item\\">2</li>
    <li class=\\"page_item\\">3</li>
    <li class=\\"current_page click_disable page_item\\">4</li>
    <li class=\\"page_item\\">5</li>
    <li class=\\"page_item\\">6</li>
    <li class=\\"\\">次へ</li>
    <li class=\\"\\">最後</li>
  </ul>
</div>"
`;

exports[`Pagination.vue currentPage=10 1`] = `
"<div class=\\"event_pagination\\">
  <ul>
    <li class=\\"\\">最初</li>
    <li class=\\"\\">前へ</li>
    <li class=\\"page_item\\">8</li>
    <li class=\\"page_item\\">9</li>
    <li class=\\"current_page click_disable page_item\\">10</li>
    <li class=\\"page_item\\">11</li>
    <li class=\\"page_item\\">12</li>
    <li class=\\"\\">次へ</li>
    <li class=\\"\\">最後</li>
  </ul>
</div>"
`;

exports[`Pagination.vue currentPage=97 1`] = `
"<div class=\\"event_pagination\\">
  <ul>
    <li class=\\"\\">最初</li>
    <li class=\\"\\">前へ</li>
    <li class=\\"page_item\\">95</li>
    <li class=\\"page_item\\">96</li>
    <li class=\\"current_page click_disable page_item\\">97</li>
    <li class=\\"page_item\\">98</li>
    <li class=\\"page_item\\">99</li>
    <li class=\\"\\">次へ</li>
    <li class=\\"\\">最後</li>
  </ul>
</div>"
`;

exports[`Pagination.vue currentPage=98 1`] = `
"<div class=\\"event_pagination\\">
  <ul>
    <li class=\\"\\">最初</li>
    <li class=\\"\\">前へ</li>
    <li class=\\"page_item\\">96</li>
    <li class=\\"page_item\\">97</li>
    <li class=\\"current_page click_disable page_item\\">98</li>
    <li class=\\"page_item\\">99</li>
    <li class=\\"page_item\\">100</li>
    <li class=\\"\\">次へ</li>
    <li class=\\"\\">最後</li>
  </ul>
</div>"
`;

exports[`Pagination.vue currentPage=99 1`] = `
"<div class=\\"event_pagination\\">
  <ul>
    <li class=\\"\\">最初</li>
    <li class=\\"\\">前へ</li>
    <li class=\\"page_item\\">97</li>
    <li class=\\"page_item\\">98</li>
    <li class=\\"current_page click_disable page_item\\">99</li>
    <li class=\\"page_item\\">100</li>
    <li class=\\"\\">次へ</li>
    <li class=\\"\\">最後</li>
  </ul>
</div>"
`;

exports[`Pagination.vue currentPage=100 1`] = `
"<div class=\\"event_pagination\\">
  <ul>
    <li class=\\"\\">最初</li>
    <li class=\\"\\">前へ</li>
    <li class=\\"page_item\\">98</li>
    <li class=\\"page_item\\">99</li>
    <li class=\\"current_page click_disable page_item\\">100</li>
    <li class=\\"click_disable\\">次へ</li>
    <li class=\\"click_disable\\">最後</li>
  </ul>
</div>"
`;
